From 76c92649444ee349499a65dc0828ea1058afe527 Mon Sep 17 00:00:00 2001
From: Anson Huang <Anson.Huang@nxp.com>
Date: Tue, 19 Jun 2018 13:27:27 +0800
Subject: [PATCH 4027/5242] MLK-18632-2 soc: imx: add support for setting
 wakeup source in ATF

commit  851a6b7a98807aeeb3d0d0d591a63d5a7c82b44d from
https://source.codeaurora.org/external/imx/linux-imx.git

To support lowest power mode for suspend, if no wakeup source
from non-secure partition is enabled, IRQSTEER can be powered
off when suspend, otherwise, IRQSTEER needs to be powered on
to support wakeup, so need to pass WU domain wakeup source
info to ATF, then ATF will decide if to power off IRQSTEER
when system suspend.

Signed-off-by: Anson Huang <Anson.Huang@nxp.com>
Reviewed-by: Bai Ping <ping.bai@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/soc/imx/pm-domains.c |   24 ++++++++++++++++++++++++
 include/soc/imx/fsl_sip.h    |    4 ++++
 2 files changed, 28 insertions(+)

diff --git a/drivers/soc/imx/pm-domains.c b/drivers/soc/imx/pm-domains.c
index 59c4cfc..bcb3632 100644
--- a/drivers/soc/imx/pm-domains.c
+++ b/drivers/soc/imx/pm-domains.c
@@ -13,6 +13,7 @@
  * GNU General Public License for more details.
  */
 
+#include <linux/arm-smccc.h>
 #include <linux/clk.h>
 #include <linux/clk-provider.h>
 #include <linux/console.h>
@@ -31,6 +32,7 @@
 #include <linux/slab.h>
 #include <linux/syscore_ops.h>
 
+#include <soc/imx/fsl_sip.h>
 #include <soc/imx8/sc/sci.h>
 
 #include "pm-domain-imx8.h"
@@ -260,6 +262,27 @@ static void imx8_detach_dev(struct generic_pm_domain *genpd, struct device *dev)
 	}
 }
 
+static int imx8_pm_domains_suspend(void)
+{
+	struct arm_smccc_res res;
+	unsigned int i;
+
+	for (i = 0; i < IMX8_WU_MAX_IRQS / 32; i++) {
+		if (wakeup_rsrc_id[i] != 0) {
+			arm_smccc_smc(FSL_SIP_WAKEUP_SRC,
+				FSL_SIP_WAKEUP_SRC_IRQSTEER, 0,
+				0, 0, 0, 0, 0, &res);
+			return 0;
+		}
+	}
+
+	arm_smccc_smc(FSL_SIP_WAKEUP_SRC,
+			FSL_SIP_WAKEUP_SRC_SCU, 0,
+			0, 0, 0, 0, 0, &res);
+
+	return 0;
+}
+
 static void imx8_pm_domains_resume(void)
 {
 	sc_err_t sci_err = SC_ERR_NONE;
@@ -278,6 +301,7 @@ static void imx8_pm_domains_resume(void)
 }
 
 struct syscore_ops imx8_pm_domains_syscore_ops = {
+	.suspend = imx8_pm_domains_suspend,
 	.resume = imx8_pm_domains_resume,
 };
 
diff --git a/include/soc/imx/fsl_sip.h b/include/soc/imx/fsl_sip.h
index a5682ca..f1c8c9b 100644
--- a/include/soc/imx/fsl_sip.h
+++ b/include/soc/imx/fsl_sip.h
@@ -60,4 +60,8 @@
 #define FSL_SIP_NOC			0xc2000008
 #define FSL_SIP_NOC_LCDIF		0x0
 
+#define FSL_SIP_WAKEUP_SRC		0xc2000009
+#define FSL_SIP_WAKEUP_SRC_SCU		0x1
+#define FSL_SIP_WAKEUP_SRC_IRQSTEER	0x2
+
 #endif
-- 
1.7.9.5

